File: day1.py
from aoc import Aoc
import itertools
import math
import re
import sys

# Day 1
# https://adventofcode.com/2021

class Day1Solution(Aoc):

    def Run(self):
        self.StartDay(1)
        self.ReadInput()
        self.PartA()
        self.PartB()

    def Test(self):
        self.StartDay(1)

        goal = self.TestDataA()
        self.PartA()
        assert self.GetAnswerA() == goal

        goal = self.TestDataB()
        self.PartB()
        assert self.GetAnswerB() == goal

    def TestDataA(self):
        self.inputdata.clear()
        testdata = \
        """
        199
        200
        208
        210
        200
        207
        240
        269
        260
        263
        """
        self.inputdata = [line.strip() for line in testdata.strip().split("\n")]
        return 7

    def TestDataB(self):
        self.inputdata.clear()
        self.TestDataA()
        return 5

    def PartA(self):
        self.StartPartA()

        numbers = [int(value) for value in self.inputdata]
        answer = len([x for x in range(1, len(numbers)) if numbers[x] > numbers[x - 1] ])

        self.ShowAnswer(answer)

    def PartB(self):
        self.StartPartB()

        numbers = [int(value) for value in self.inputdata]
        groups = [sum(numbers[x:x + 3]) for x in range(0, len(numbers)) if x < len(numbers) - 2]
        answer = len([x for x in range(1, len(groups)) if groups[x] > groups[x - 1] ])

        self.ShowAnswer(answer)


if __name__ == "__main__":
    solution = Day1Solution()
    if len(sys.argv) >= 2 and sys.argv[1] == "test":
        solution.Test()
    else:
        solution.Run()

File: c/day1.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "aoc.h"

#define DAY 1

#define MAX_LINES 3000
#define MAX_LINE_LENGTH 10

void part_a();
void part_b();

char** lines;
int line_count = 0;
int* numbers;

void main()
{
	start_day(DAY);

	lines = read_input(DAY, "../", MAX_LINES, MAX_LINE_LENGTH, &line_count);
	if (lines == NULL)
	{
		perror("read_input()");
		return;
	}

    numbers = convert_input_to_numbers(lines, line_count);

    part_a();
    part_b();

    free(numbers);
	free_input(lines, line_count);
}

void part_a()
{
	start_part_a();

	int increase_count = 0;
	for (int ix = 1; ix < line_count; ix++)
	{
		if (numbers[ix] > numbers[ix - 1])
		{
			increase_count++;
		}
	}

	show_answer(increase_count);
}

void part_b()
{
	start_part_b();

	int increase_count = 0;
	int prev_sum = -1;
	for (int ix = 0; ix < line_count - 2; ix++)
	{
		int sum = numbers[ix] + numbers[ix + 1] + numbers[ix + 2];
		if (sum > prev_sum && prev_sum != -1)
		{
			increase_count++;
		}

		prev_sum = sum;
	}

	show_answer(increase_count);
}
